File: sdks/cpp/direct.cpp
// direct.cpp — patala from C++, in-process, through the C ABI.
//
// The same payout `sdks/c/direct.c` walks, written the way C++ writes it: the
// handle and every returned string are owned by objects, so there is no
// cleanup label, no `goto`, and no path — including a throw — that leaks.
//
// This is an EXAMPLE, not a test. The ABI's test is
// patala-ffi/ctest/smoke.c, which dlopens the artifact and resolves every
// symbol by name.
//
// Everything runs on MockRail: deterministic, offline, no credentials, no
// network. This is a payments library, and an example that moves real value is
// not an example.
//
// SPDX-License-Identifier: MIT OR Apache-2.0

#include <cstdio>
#include <iostream>
#include <string>

#include "jsonpeek.hpp"
#include "patala.hpp"

#ifndef PATALA_EXPECTED_VERSION
#define PATALA_EXPECTED_VERSION "0.0.0-unset"
#endif

namespace {

// Spelled out rather than defaulted so the shape of a real configuration is
// visible. Unknown fields are refused, so a typo throws.
const char *kRailConfig = R"({"rail":"mock","currencies":["USDC","USD"]})";

// What the customer typed into the payout form.
const char *kCustomerAddress = "mock:wallet:alice";

const char *kPayRequest =
    R"({"amount_minor":1250,"currency":"USDC","destination":"mock:wallet:alice","reference":"order-1"})";

} // namespace

int main() {
	try {
		// --- 0. the version probe ---------------------------------------
		std::cout << "patala direct (C++, in-process) — libpatala_ffi " << patala::abi_version()
		          << "\n";
		patala::require_abi(PATALA_EXPECTED_VERSION);
		std::cout << "abi:       matches " PATALA_EXPECTED_VERSION "\n";

		// --- 1. open a rail ----------------------------------------------
		// The destructor closes it. There is no close() to forget, and an
		// exception thrown anywhere below still runs it.
		const patala::Rail rail{kRailConfig};
		std::cout << "rail:      handle " << rail.handle() << "\n";

		// --- 2. capabilities decide the UX -------------------------------
		// Not a bool: "CustodialReversible" means a card form and a refundable
		// pending state, "NonCustodialFinal" means a wallet address and a
		// signed final receipt. Those are not two shades of one thing.
		const std::string caps = rail.capabilities();
		const std::string rail_class = peek::str(caps, "class").value_or("?");
		std::cout << "caps:      " << rail_class << " / "
		          << (rail_class == "CustodialReversible" ? "card form, refundable pending state"
		                                                  : "wallet address, signed final receipt")
		          << "\n";
		std::cout << "           holds_funds="
		          << (peek::flag(caps, "holds_funds").value_or(true) ? "true" : "false")
		          << " — patala itself never holds funds\n";

		// --- 3. the caveat, before there is a verdict ---------------------
		// The wording to show on the form where the address is first asked
		// for. It is its own method because at that moment there is nothing
		// to validate yet.
		const std::string caveat =
		    peek::str(rail.caveat(), "exchange_deposit_caveat").value_or("");
		std::cout << "caveat:    " << caveat.substr(0, 72) << "...\n";

		// --- 4. the pre-flight check ---------------------------------------
		// Offline, and it NEVER fails: "I cannot check this" comes back as the
		// verdict Unknown, not as an exception, because a caller must handle
		// it as carefully as a refusal.
		const std::string verdict =
		    rail.validate_destination(std::string(R"({"destination":")") + kCustomerAddress + "\"}");
		const bool refused = peek::flag(verdict, "is_refusal").value_or(true);
		std::cout << "dest:      " << kCustomerAddress << " -> "
		          << peek::str(verdict, "status").value_or("?")
		          << " (is_refusal=" << (refused ? "true" : "false") << ", human_must_confirm="
		          << (peek::flag(verdict, "human_must_confirm").value_or(false) ? "true" : "false")
		          << ")\n";
		if (refused) {
			std::cout << "           refused — do not send. Stop here.\n";
			return 1;
		}
		// human_must_confirm is true on EVERY verdict, including
		// StructurallyValid. patala does not detect exchange-owned addresses
		// and will not guess, so a person still has to say yes.

		// --- 5. quote ------------------------------------------------------
		const std::string quote = rail.quote(kPayRequest);
		std::cout << "quote:     " << peek::u64(quote, "amount_minor").value_or(0) << " + "
		          << peek::u64(quote, "fee_minor").value_or(0) << " fee = "
		          << peek::u64(quote, "total_minor").value_or(0) << " minor units of "
		          << peek::str(quote, "currency").value_or("?") << "\n";

		// --- 6. charge -------------------------------------------------------
		// The Receipt is the entitlement. Store the whole document; it is what
		// you hand back to verify() later.
		const std::string receipt = rail.charge(kPayRequest);
		std::cout << "charge:    " << peek::u64(receipt, "amount_minor").value_or(0)
		          << " minor units, ref=" << peek::str(receipt, "reference").value_or("?")
		          << ", issued by rail " << peek::str(receipt, "rail_id").value_or("?") << "\n";

		// --- 7. verify --------------------------------------------------------
		// Gate entitlement on this, never on charge() having returned.
		std::cout << "verify:    " << rail.verify(receipt) << "  <- the entitlement check\n";

		// --- 8. the outcome that is NOT an exception ---------------------------
		// verify_holds() returns false for a tampered receipt and THROWS only
		// when the check could not be performed. Merging the two is how an
		// unpaid order becomes an entitlement: every `catch` that logs and
		// retries would be granting one.
		std::string tampered = receipt;
		const std::string needle = "\"amount_minor\":1250";
		const std::size_t at = tampered.find(needle);
		if (at != std::string::npos) {
			tampered.replace(at, needle.size(), "\"amount_minor\":999999");
			std::cout << "tampered:  " << rail.verify(tampered) << " -> verify_holds()="
			          << (rail.verify_holds(tampered) ? "true (WRONG)" : "false")
			          << "  <- returned, not thrown\n";
		}

		// --- 9. the throw path, and what it does not leak ----------------------
		// An operational refusal IS an exception, and the library's message is
		// malloc'd on the other side of the ABI. `detail::Owned` holds it
		// across the std::string copy and frees it during unwinding — the one
		// line a hand-written binding forgets. `leaks --atExit -- ./direct`
		// reports zero; see README.md.
		try {
			rail.charge(
			    R"({"amount_minor":100,"currency":"EUR","destination":"mock:wallet:alice","reference":"order-2"})");
			std::cout << "refused:   (a receipt — WRONG)\n";
		} catch (const patala::Error &e) {
			std::cout << "refused:   patala::Error: " << e.what() << "\n";
		}

		// A closed or invented handle is a clean error, never a segfault in
		// your address space: handles are registry keys, and they are never
		// reused.
		try {
			patala::Rail moved{kRailConfig};
			const patala::Rail took = std::move(moved); // moved-from handle is 0
			std::cout << "moved:     handle " << moved.handle() << " -> " << took.handle()
			          << ", one owner, one close\n";
		} catch (const patala::Error &e) {
			std::cout << "moved:     unexpected: " << e.what() << "\n";
		}

		std::cout << "\nOK — offline, MockRail only, no value moved.\n";
		return 0;
	} catch (const patala::Error &e) {
		std::cerr << "patala::Error: " << e.what() << "\n";
		return 1;
	} catch (const std::exception &e) {
		std::cerr << "error: " << e.what() << "\n";
		return 1;
	}
}

File: sdks/node/direct.ts
// patala, direct — libpatala_ffi loaded into your Node process over the C ABI
// in patala-ffi/include/patala.h.
//
//   import { Rail } from "patala/direct";
//
//   using rail = Rail.open({ rail: "mock", currencies: ["USDC"] });
//   const receipt = rail.charge({ amount_minor: 1250, currency: "USDC",
//                                 destination: "mock:wallet:alice", reference: "order-1" });
//   rail.verify(receipt).valid;                       // true — THIS is the entitlement
//
// WHAT LOADING THIS COSTS YOUR PROCESS: measurably nothing. patala is Rust, so
// there is no language runtime in the host — no GC, no scheduler, no signal
// handlers replaced, no threads started. Measured from Node on this machine,
// 7 threads before dlopen and 7 after a full charge -> verify round trip; the
// Go-based C ABI in the sibling product openrate takes the same process from 7
// to 13. README.md has the table, and examples/direct.ts prints the count.
//
// The practical consequence, and it is a big one: **Node worker threads and
// koffi's off-thread `.async` path both work here and both let the process
// exit.** Neither does against llmux's or openrate's Go libraries. So unlike
// those two SDKs, this one is not condemned to be synchronous — see
// `callAsync` below and README.md's "Off the main thread".
//
// There is deliberately no streaming entry point, here or in the ABI: patala
// has no streaming operation. Nothing it does produces a sequence of chunks, so
// there is nothing to iterate. (llmux, which shares this ABI shape, does have
// `llmux_stream`. Do not go looking for patala's.)
//
// JSON in, JSON out — the same JSON patala-sidecar serves.

import * as fs from "fs";
import * as path from "path";
import type { KoffiFunc, LibraryHandle } from "koffi";

import type {
  CaveatResult,
  DestinationVerdict,
  IdResult,
  PayRequest,
  ProvidersResult,
  Quote,
  RailCapabilities,
  RailConfig,
  Receipt,
  ResultOf,
  VerifyResult,
  WebhookDelivery,
  WebhookEvent,
} from "./types.js";
import { narrowVerdict, narrowVerify } from "./types.js";

export * from "./types.js";

// ---------------------------------------------------------------------------
// koffi, loaded lazily
// ---------------------------------------------------------------------------
// koffi is an OPTIONAL dependency, so `require("patala")` — the sidecar path,
// which needs no native code at all — works without it. See README.md for why
// koffi and not a hand-written N-API addon.

/** Anything koffi hands back that is a raw C pointer. Never dereference it in JS. */
type CPtr = unknown;
/** koffi's `_Out_ void**` convention: a one-element array koffi writes into. */
type ErrOut = [CPtr];

interface Koffi {
  load(filename: string): LibraryHandle;
  decode(value: CPtr, type: string, len: number): unknown;
}

let _koffi: Koffi | null = null;
function koffi(): Koffi {
  if (_koffi) return _koffi;
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports -- optional dependency, resolved at first direct-mode use
    _koffi = require("koffi") as Koffi;
  } catch {
    throw new Error(
      'patala direct mode needs the optional "koffi" dependency: npm install koffi. ' +
        'Or use the sidecar — `require("patala").Sidecar.start()` — which needs no native FFI.'
    );
  }
  return _koffi;
}

// ---------------------------------------------------------------------------
// Finding libpatala_ffi
// ---------------------------------------------------------------------------

function libFileName(): string {
  // cargo names the cdylib libpatala_ffi.{dylib,so} / patala_ffi.dll.
  if (process.platform === "darwin") return "libpatala_ffi.dylib";
  if (process.platform === "win32") return "patala_ffi.dll";
  return "libpatala_ffi.so";
}

/**
 * Where the shared library will be loaded from.
 *
 * `PATALA_LIBRARY` wins. Otherwise a repo checkout's `target/release/` is
 * tried, then `target/debug/`, and failing both the bare name goes to the
 * system loader. Release first because that is the artifact whose size patala
 * advertises; if you built only a debug one, that is what you get.
 *
 * A stale library earlier on the load path is the classic way for a binding to
 * misbehave in ways that look like patala bugs, which is why the ABI carries
 * `patala_abi_check` — pass `expectVersion` to {@link Rail.open} and the
 * library itself does the comparison.
 *
 * **Built and executed here: darwin/arm64 only.** See README.md; nothing in
 * this package implies a Linux .so or a Windows .dll exists.
 */
export function resolveLibrary(explicit?: string): string {
  if (explicit) return explicit;
  if (process.env.PATALA_LIBRARY) return process.env.PATALA_LIBRARY;
  const root = path.join(__dirname, "..", "..");
  for (const profile of ["release", "debug"]) {
    const candidate = path.join(root, "target", profile, libFileName());
    if (fs.existsSync(candidate)) return candidate;
  }
  return libFileName();
}

// ---------------------------------------------------------------------------
// The six symbols
// ---------------------------------------------------------------------------

/**
 * koffi decorates every bound function with an off-thread `.async` variant that
 * takes a Node-style callback. Its own typings do not model it, so this does.
 */
type AsyncCapable<Args extends unknown[], R> = KoffiFunc<(...args: Args) => R> & {
  async(...args: [...Args, (err: Error | null, value: R) => void]): void;
};

interface Binding {
  path: string;
  abi_version: KoffiFunc<() => string>;
  abi_check: KoffiFunc<(expected: string, err: ErrOut) => number>;
  new_: KoffiFunc<(config: string | null, err: ErrOut) => number>;
  call: AsyncCapable<[h: number, method: string, req: string | null, err: ErrOut], CPtr>;
  close: KoffiFunc<(h: number) => void>;
  free: KoffiFunc<(p: CPtr) => void>;
}

const _bindings = new Map<string, Binding>();

function bind(libPath: string): Binding {
  const cached = _bindings.get(libPath);
  if (cached) return cached;
  const lib = koffi().load(libPath);
  const b: Binding = {
    path: libPath,
    // `const char*`: koffi decodes it and does not free it. Correct here and
    // only here — patala_abi_version returns a static string the library owns.
    abi_version: lib.func("const char *patala_abi_version()"),
    abi_check: lib.func("int patala_abi_check(const char *expected, _Out_ void **err)"),
    new_: lib.func("uint64_t patala_new(const char *config_json, _Out_ void **err)"),
    // Declared `void*`, not `char*`, ON PURPOSE: koffi would decode a `char*`
    // result into a JS string and drop the pointer, leaving nothing to hand
    // patala_free. That is the leak this line prevents.
    call: lib.func(
      "void *patala_call(uint64_t h, const char *method, const char *request_json, _Out_ void **err)"
    ) as Binding["call"],
    close: lib.func("void patala_close(uint64_t h)"),
    free: lib.func("void patala_free(void *p)"),
  };
  _bindings.set(libPath, b);
  return b;
}

/** The patala version the loaded shared library was built from. */
export function abiVersion(libraryPath?: string): string {
  return bind(resolveLibrary(libraryPath)).abi_version();
}

/**
 * Ask the library whether it is the version you generated against, and throw
 * its own message if not.
 *
 * This is `patala_abi_check`, not a comparison written here. The ABI exports it
 * precisely so twelve bindings do not each reimplement — and each forget — the
 * check.
 */
export function abiCheck(expected: string, libraryPath?: string): void {
  const b = bind(resolveLibrary(libraryPath));
  const err: ErrOut = [null];
  if (b.abi_check(expected, err) !== 0) throw takeError(b, err, `patala_abi_check(${expected}) failed`);
}

/** Read a C string patala allocated, then free it. Freeing is not optional. */
function takeString(b: Binding, p: CPtr): string | null {
  if (!p) return null;
  try {
    return koffi().decode(p, "char", -1) as string;
  } finally {
    b.free(p);
  }
}

/** Turn a populated `char** err` into an Error, freeing the message. */
function takeError(b: Binding, err: ErrOut, fallback: string): Error {
  // Error strings are plain UTF-8 text, NOT JSON. Do not parse them.
  const msg = takeString(b, err[0]);
  err[0] = null;
  return new Error(msg ?? fallback);
}

function encodeRequest(request: unknown): string | null {
  if (request == null) return null;
  return typeof request === "string" ? request : JSON.stringify(request);
}

// ---------------------------------------------------------------------------
// Rail
// ---------------------------------------------------------------------------

export interface RailOptions {
  /** Override the shared library path (otherwise {@link resolveLibrary}). */
  libraryPath?: string;
  /**
   * Refuse to open unless the loaded library reports this version. Checked by
   * `patala_abi_check` inside the library, not by a comparison here.
   */
  expectVersion?: string;
}

/**
 * One rail, behind one handle.
 *
 * Opening a rail talks to nothing: no socket, no thread, no file. Only
 * {@link call} reaches a network, and only for a rail that has one — the
 * default `mock` rail has none at all, which is why every example in this
 * package uses it. This is a payments library; an example that moves real
 * value is not an example.
 *
 * Disposable, so `using rail = Rail.open()` closes the handle on every exit
 * path out of the block, including a throw.
 */
export class Rail implements Disposable {
  private readonly b: Binding;
  private h: number;
  private closed = false;

  private constructor(b: Binding, h: number) {
    this.b = b;
    this.h = h;
  }

  /**
   * Build a rail. `undefined` means the offline default — a deterministic
   * `MockRail` on USDC, needing no credentials and no network.
   */
  static open(config?: RailConfig, opts: RailOptions = {}): Rail {
    const libPath = resolveLibrary(opts.libraryPath);
    const b = bind(libPath);
    if (opts.expectVersion !== undefined) abiCheck(opts.expectVersion, libPath);
    const err: ErrOut = [null];
    const h = b.new_(config === undefined ? null : JSON.stringify(config), err);
    // 0 is patala_new's failure sentinel, because its success value is a
    // handle and handles start at 1.
    if (h === 0) throw takeError(b, err, "patala_new failed");
    return new Rail(b, h);
  }

  /** The registry key of this rail. Handles are retired, never recycled. */
  get handle(): number {
    return this.h;
  }

  private live(): number {
    if (this.closed) throw new Error("patala rail is closed");
    return this.h;
  }

  /**
   * Any method, typed by name. The nine are `id`, `capabilities`, `quote`,
   * `charge`, `verify`, `validate-destination`, `webhook`, `caveat` and
   * `providers`; the named methods below are thin wrappers over this.
   */
  call<M extends keyof ResultOf>(method: M, request?: unknown): ResultOf[M] {
    const h = this.live();
    const err: ErrOut = [null];
    const res = this.b.call(h, method, encodeRequest(request), err);
    if (!res) throw takeError(this.b, err, `patala_call(${method}) failed`);
    return JSON.parse(takeString(this.b, res) ?? "null") as ResultOf[M];
  }

  /**
   * The same call, run on a libuv threadpool thread instead of this one.
   *
   * Pointless for the mock rail, which answers in microseconds. Not pointless
   * for a real one: `charge` on Solana, Stellar or a fiat processor is a
   * network round trip, and a synchronous FFI call would block Node's event
   * loop for its whole duration.
   *
   * **That this works at all is patala-specific.** The identical construction
   * against llmux's or openrate's Go libraries returns the right answer and
   * then hangs the process forever, which is why both of those SDKs are
   * synchronous-only. patala has no runtime in the host, so there is no thread
   * stuck inside one at exit. Verified, with the failing control, in README.md.
   */
  async callAsync<M extends keyof ResultOf>(method: M, request?: unknown): Promise<ResultOf[M]> {
    const h = this.live();
    const err: ErrOut = [null];
    const body = encodeRequest(request);
    const res = await new Promise<CPtr>((resolve, reject) => {
      this.b.call.async(h, method, body, err, (e: Error | null, value: CPtr) => {
        if (e) reject(e);
        else resolve(value);
      });
    });
    if (!res) throw takeError(this.b, err, `patala_call(${method}) failed`);
    return JSON.parse(takeString(this.b, res) ?? "null") as ResultOf[M];
  }

  /** This rail's stable id — `"mock"`, `"solana"`, … */
  id(): IdResult {
    return this.call("id");
  }

  /**
   * What this rail is and is not able to do. Decide your whole UX from
   * `class` without knowing which provider answered.
   */
  capabilities(): RailCapabilities {
    return this.call("capabilities");
  }

  /** Fees, fx and expiry. Moves no money. */
  quote(req: PayRequest): Quote {
    return this.call("quote", req);
  }

  /**
   * Move money and return the receipt. **Store the receipt** — it, not this
   * call returning, is the entitlement.
   */
  charge(req: PayRequest): Receipt {
    return this.call("charge", req);
  }

  /**
   * Re-derive whether a receipt still holds. `{valid: false}` is a RESULT, not
   * an error: it is the rail's fail-closed answer. Gate on `valid === true` and
   * nothing else, and never retry a `false` as though it were transient.
   */
  verify(receipt: Receipt): VerifyResult {
    return narrowVerify(this.call("verify", receipt));
  }

  /**
   * The offline pre-flight check to run before any money moves. It never
   * fails — "I cannot check this" comes back as `status: "Unknown"`. Read
   * `is_refusal` (do not send) and `human_must_confirm`, which is `true` on
   * every verdict including `StructurallyValid`.
   */
  validateDestination(destination: string): DestinationVerdict {
    return narrowVerdict(this.call("validate-destination", { destination }));
  }

  /**
   * Authenticate an inbound delivery from a rail's processor. Forward it
   * VERBATIM: every scheme signs the exact bytes that were sent, so a body
   * that has been through a JSON round-trip on your side no longer matches its
   * own signature.
   *
   * A rail with no push delivery — the mock, fiat's `manual` — refuses rather
   * than inventing an event.
   */
  webhook(delivery: WebhookDelivery): WebhookEvent {
    return this.call("webhook", delivery);
  }

  /** The sentence to show a human before they confirm a payout address. */
  caveat(): CaveatResult {
    return this.call("caveat");
  }

  /** Every fiat provider this build has compiled in. Fiat builds only. */
  providers(): ProvidersResult {
    return this.call("providers");
  }

  /**
   * Release the rail. Idempotent, as `patala_close` is — closing an unknown or
   * already-closed handle is a no-op, so a cleanup path can be unconditional.
   */
  close(): void {
    if (this.closed) return;
    this.closed = true;
    this.b.close(this.h);
  }

  [Symbol.dispose](): void {
    this.close();
  }
}
